pyrenew_hew/ww_site_level_dynamics_model.py

```diff
@@ -0,0 +1,570 @@
+            ):  # only fit hosp admissions data
+                n_datapoints = len(data_observed_hospital_admissions)
+            else:  # both hosp admisssions and ww data provided
+                n_datapoints = max(
+                    len(data_observed_hospital_admissions),
+                    self.max_ww_sampled_days,
+                )
+        else:
+            if (
+                data_observed_hospital_admissions is not None
+                or data_observed_log_conc is not None
+            ):
+                raise ValueError(
+                    "Cannot pass both n_datapoints and "
+                    "data_observed_hospital_admissions "
+                    "or data_observed_log_conc"
+                )
+            else:
+                n_datapoints = n_datapoints
+
+        n_weeks_post_init = -((-n_datapoints) // 7)  # n_datapoints // 7 + 1
+
+        eta_sd = self.eta_sd_rv()
+        autoreg_rt = self.autoreg_rt_rv()
+        log_r_t_first_obs = self.log_r_t_first_obs_rv()
+
+        rt_init_rate_of_change_rv = DistributionalVariable(
+            "rt_init_rate_of_change",
+            dist.Normal(0, eta_sd / jnp.sqrt(1 - jnp.pow(autoreg_rt, 2))),
+        )
+
+        rt_init_rate_of_change = rt_init_rate_of_change_rv()
+
+        log_r_t_in_weeks = self.ar_diff_rt(
+            noise_name="rtu_weekly_diff_first_diff_ar_process_noise",
+            n=n_weeks_post_init,
+            init_vals=jnp.array(log_r_t_first_obs),
+            autoreg=jnp.array(autoreg_rt),
+            noise_sd=jnp.array(eta_sd),
+            fundamental_process_init_vals=jnp.array(rt_init_rate_of_change),
+        )
+        numpyro.deterministic("log_r_t_in_weeks", log_r_t_in_weeks)
+
+        i_first_obs_over_n = self.i_first_obs_over_n_rv()
+        offset_ref_logit_i_first_obs = self.offset_ref_logit_i_first_obs_rv()
+
+        mean_initial_exp_growth_rate = self.mean_initial_exp_growth_rate_rv()
+        offset_ref_initial_exp_growth_rate = (
+            self.offset_ref_initial_exp_growth_rate_rv()
+        )
+
+        i_first_obs_over_n_ref_subpop = transforms.SigmoidTransform()(
+            transforms.logit(i_first_obs_over_n)
+            + jnp.where(self.n_subpops > 1, offset_ref_logit_i_first_obs, 0)
+        )
+        initial_exp_growth_rate_ref_subpop = (
+            mean_initial_exp_growth_rate
+            + jnp.where(
+                self.n_subpops > 1, offset_ref_initial_exp_growth_rate, 0
+            )
+        )
+
+        offset_ref_log_r_t = self.offset_ref_log_r_t_rv()
+        log_rtu_ref_subpop_in_week = log_r_t_in_weeks + jnp.where(
+            self.n_subpops > 1, offset_ref_log_r_t, 0
+        )
+
+        if self.n_subpops == 1:
+            i_first_obs_over_n_subpop = i_first_obs_over_n_ref_subpop
+            initial_exp_growth_rate_subpop = initial_exp_growth_rate_ref_subpop
+            log_rtu_subpop_in_week = log_rtu_ref_subpop_in_week[:, jnp.newaxis]
+        else:
```

Comment:
For reviewing and onboarding people to the code, I think it would be helpful if there were a `.md` file or more comments throughout the code to explain what is going on and how data is structured.

E.g. if there is only 1 subpop, `i_first_obs_over_n_subpop` has shape x, but if there is more than one, it has shape y, where the dimension represent a and b`

Reply:
For instance, I am curious why `log_rtu_subpop_in_week` requires `jnp.newaxis`, but `i_first_obs_over_n_subpop` and `initial_exp_growth_rate_subpop` do not.

Reply:
Sometimes I used `jnp.newaxis` to expand 1D array to ensure matrix operations work as intended.

In the part above, I extend `log_rtu_subpop_in_week` so that the code below works both when `n_subpops =1` and `n_subpops >1` rather than having separate codes for either case. But `initial_exp_growth_rate_subpop` works for either case without the need to add additional axis.
```
rtu_subpop = jnp.squeeze(
            jnp.repeat(
                jnp.exp(log_rtu_subpop_in_week),
                repeats=7,
                axis=0,
            )[:n_datapoints, :]
        )
```